cpp/include/kvikio/shim/libcurl.hpp

```diff
@@ -0,0 +1,260 @@
+ * to think about it. (Caveat: If you are initializing libcurl from a Windows DLL you should not
+ * initialize it from DllMain or a static initializer because Windows holds the loader lock during
+ * that time and it could cause a deadlock.)
+ *
+ * Source <https://curl.se/libcurl/c/libcurl.html>.
+ */
+class LibCurl {
+ public:
+  // We hold an unique pointer to the raw curl handle and sets `curl_easy_cleanup` as its Deleter.
+  using UniqueHandlePtr = std::unique_ptr<CURL, std::function<void(CURL*)>>;
+
+ private:
+  std::mutex _mutex{};
+  // Curl handles free to be used.
+  std::vector<UniqueHandlePtr> _free_curl_handles{};
+
+  LibCurl()
+  {
+    CURLcode err = curl_global_init(CURL_GLOBAL_DEFAULT);
+    if (err != CURLE_OK) {
+      throw std::runtime_error("cannot initialize libcurl - errorcode: " + std::to_string(err));
+    }
+    curl_version_info_data* ver = curl_version_info(::CURLVERSION_NOW);
+    if ((ver->features & CURL_VERSION_THREADSAFE) == 0) {
+      throw std::runtime_error("cannot initialize libcurl - built with thread safety disabled");
+    }
+  }
+  ~LibCurl() noexcept
+  {
+    _free_curl_handles.clear();
+    curl_global_cleanup();
+  }
+
+ public:
+  static LibCurl& instance()
+  {
+    static LibCurl _instance;
+    return _instance;
+  }
+
+  /**
+   * @brief Returns a free curl handle if available.
+   */
+  UniqueHandlePtr get_free_handle()
+  {
+    UniqueHandlePtr ret;
+    std::lock_guard const lock(_mutex);
+    if (!_free_curl_handles.empty()) {
+      ret = std::move(_free_curl_handles.back());
+      _free_curl_handles.pop_back();
+    }
+    return ret;
+  }
+
+  /**
+   * @brief Returns a curl handle, create a new handle if none is available.
+   */
+  UniqueHandlePtr get_handle()
+  {
+    // Check if we have a free handle available.
+    UniqueHandlePtr ret = get_free_handle();
+    if (ret) {
+      curl_easy_reset(ret.get());
+    } else {
+      // If not, we create a new handle.
+      CURL* raw_handle = curl_easy_init();
+      if (raw_handle == nullptr) {
+        throw std::runtime_error("libcurl: call to curl_easy_init() failed");
+      }
+      ret = UniqueHandlePtr(raw_handle, curl_easy_cleanup);
+    }
+    return ret;
+  }
+
+  /**
+   * @brief Retain a curl handle for later use.
+   */
+  void retain_handle(UniqueHandlePtr handle)
+  {
+    std::lock_guard const lock(_mutex);
+    _free_curl_handles.push_back(std::move(handle));
+  }
+};
+
+/**
+ * @brief Representation of a curl easy handle pointer and its operations.
+ *
+ * An instance is given a `LibCurl::UniqueHandlePtr` on creation, which is
+ * later retailed on destruction.
+ */
+class CurlHandle {
+ private:
+  char _errbuf[CURL_ERROR_SIZE];
+  LibCurl::UniqueHandlePtr _handle;
+  std::string _source_file;
+  std::string _source_line;
+
+ public:
+  /**
+   * @brief Construct a new curl handle.
+   *
+   * Typically, do not use this directly instead use the `create_curl_handle()` macro.
```

Comment:
Should it ever be the case that one constructs a `CurlHandle` directly?

Reply:
I guess if you want to specify some other source and/or line number?